GH2691: Make XmlnsDefinitionAttribute public

legistek opened the pull request:
### Description of Change ###

This makes `XmlnsDefinitionAttribute` `public` and enables it to be used with any assembly in the project to map C# namespaces to custom URLs.

### Bugs Fixed ###

Fixes #2691
Also resolves what may be an un-logged bug preventing using `x:Name` with elements referenced with `using:` style namespace definitions.

### API Changes ###

Added:
 None

Changed:
 - `XmlnsDefinitionAttribute` to the `Xamarin.Forms.Xaml` namespace and the `public` accessibility modifier (was `internal`)
 - XamlG, XamlC, and XamlParser have been changed to look for all referenced assemblies with the `XmlnsDefinitionAttribute` as opposed to only looking in `Xamarin.Forms.Core` and `Xamarin.Forms.Xaml`. Specifically:

1. The XamlG build task now takes a `References` input from MSBuild, consisting of a semicolon-separated list of paths to referenced assemblies in the project, and provides this to `XamlGenerator`. `XamlGenerator` now looks for `XmlnsDefinitionAttribute` in these assemblies, and then resolves type references using Mono APIs. Several members had to be changed from `static` in order to have access to the referenced assembly list. XamlG also checks for `using:` at the beginning of namespace URIs and handles these correctly.
2. The `XmlTypeExtensions.GatherXmlnsDefinitionAttributes` method used by the XamlC build task now uses the `ModuleDefinition` being compiled to look for referenced assemblies. In addition, rather than maintaining a single static list of `XmlnsDefinitionAttribute`s, a `Dictionary` is used to maintain individual lists for each `ModuleDefinition`, and the procedure is thread-safe. This seemed necessary in case there are multiple concurrent compilations occurring under the same XamlC process.
3. The `XamlParser` looks for the `XmlnsDefinitionAttribute` in all assemblies in the current `AppDomain`. 

### Behavioral Changes ###

The `XmlnsDefinitionAttribute` can now be applied to any assembly with CLR namespaces mapped to any custom URL namespace. The namespaces are recognized for all purposes: code-behind generation, XAML compiling, runtime XAML parsing, and IntelliSense. 

Note that for the behavior to work correctly on all platforms, the `Preserve` attribute must be applied to any classes in the assembly that will fall under the custom URL scheme,, if not the entire assembly. In addition, the assembly with the custom-mapped classes must be referenced in the actual code of the assembly that contains the XAML referencing the custom URL. This can be accomplished by having the consuming assembly call a simple, empty `Init` method in the control library, similar to what is necessary when making custom platform renderers. 

Finally, when making unit tests, I discovered that when employing the `using:` syntax to specify a namespace URI, it was not possible to add an `x:Name` to these elements because XamlG was not checking for `using:` at the beginning of the URI. This has been fixed.

### PR Checklist ###

- [X] Has tests (if omitted, state reason in description)
- [X] Rebased on top of the target branch at time of PR
- [X] Changes adhere to coding standard
- [X] Consolidate commits as makes sense

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=2782) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=2782) before we can accept your contribution.<br/><br/>:x: legistek [sign now](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=2782)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/xamarin/Xamarin.Forms?pullRequest=2782) it.</sub>

dansiegel commented:
just want to refresh the conversation around this... @legistek looks like it will need a rebase to resolve some conflicts... @StephaneDelcroix based on the original target we were looking at 3.2... any chance of that still happening?

legistek commented:
Hi @dansiegel  - happy to do the rebase and conflict resolution. Was just waiting to see if you guys were close to doing a review before I did so in case there were other intervening changes .

rmarinho commented:
Hey @legistek can you rebase so we can review? 

At this time, if this is pointing to master it will target in a 3.3.0 release not 3.2.0.

Thanks

legistek commented:
Sure thing @rmarinho - it's still pointing to 3.1 I believe:
https://github.com/legistek/Xamarin.Forms/commits/gh2691-xmlnsdefinitionattribute-public

so I can still rebase onto 3.2, or current master. What's your preference?

rmarinho commented:
At this time I think is best we target master.

legistek commented:
@rmarinho should be good to go now

legistek commented:
By the way, is it normal that I can't run any of the Windows UI unit tests? I can't run them from master either; always get runtime errors.

legistek commented:
Hi @rmarinho, @dansiegel  just a friendly reminder this should be ready to roll.

dansiegel commented:
@legistek I have no authority to do anything... but I'd certainly love to see this merged. It's a shame that this PR keeps going stale. I wish this had been merged months ago. I would love to use this for Prism. @davidortinau will attest to the fact that I keep bugging him to get this merged.

legistek commented:
Oh, hah sorry @dansiegel I assumed you were one of TPTB. :) Well it does touch a lot of things so I'm not surprised it would take awhile for the team to get to it.

So @rmarinho it looks like an automated build is failing on VSTS, but I don't have access to see the details. I'm not seeing any trouble building locally. Can you let me know what's actually failing so I can address if necessary? Thanks!

rmarinho commented:
@legistek 
```
Error CS0618: 'Fragment' is obsolete: 'This class is obsoleted in this android platform'
```

legistek commented:
> @legistek needs rebase
> 
> ```
> Error CS0618: 'Fragment' is obsolete: 'This class is obsoleted in this android platform'
> ```

Got it thanks, I will take care of it.

legistek commented:
@rmarinho thanks it's been successfully re-based again after a little git drama.

rmarinho commented:
build

legistek commented:
@rmarinho what's the OSX build error?

PureWeen commented:
@legistek 

```
2018-10-05T22:50:14.2592450Z Properties/AssemblyInfo.cs(5,12): error CS0122: 'XmlnsDefinitionAttribute' is inaccessible due to its protection level [/Users/vsts/agent/2.140.2/work/1/s/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.CustomNamespaceUris/Xamarin.Forms.Controls.Issues.CustomNamespaceUris.csproj]
2018-10-05T22:50:14.2596770Z Properties/AssemblyInfo.cs(6,12): error CS0122: 'XmlnsDefinitionAttribute' is inaccessible due to its protection level [/Users/vsts/agent/2.140.2/work/1/s/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.CustomNamespaceUris/Xamarin.Forms.Controls.Issues.CustomNamespaceUris.csproj]
2018-10-05T22:50:14.2598200Z Properties/AssemblyInfo.cs(7,12): error CS0122: 'XmlnsDefinitionAttribute' is inaccessible due to its protection level [/Users/vsts/agent/2.140.2/work/1/s/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.CustomNamespaceUris/Xamarin.Forms.Controls.Issues.CustomNamespaceUris.csproj]
2018-10-05T22:50:14.2760130Z Done Building Project "/Users/vsts/agent/2.140.2/work/1/s/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.CustomNamespaceUris/Xamarin.Forms.Controls.Issues.CustomNamespaceUris.csproj" (default targets) -- FAILED.
```

StephaneDelcroix commented:
> Also resolves what may be an un-logged bug preventing using x:Name with elements referenced with using: style namespace definitions.

this is, iirc, already solved elsewhere

legistek commented:
@StephaneDelcroix rebase onto master right?

legistek commented on Xamarin.Forms.Xaml.UnitTests/MSBuild/MSBuildTests.cs:
```diff
@@ -67,7 +67,7 @@ public void SetUp ()
 
 			//We need to copy Xamarin.Forms.targets to the test directory, to reliably import them
-			var xamarinFormsTargets = Path.Combine (testDirectory, "..", "..", "..", ".nuspec", XamarinFormsTargets);
+			var xamarinFormsTargets = Path.Combine (testDirectory, "..", "..", "..", "..", ".nuspec", XamarinFormsTargets);
```
FYI I'm not sure if it was just me but I couldn't get these unit tests to pass without this change.

legistek commented on Xamarin.Forms.Xaml/XamlParser.cs:
```diff
@@ -329,32 +333,119 @@ static void GatherXmlnsDefinitionAttributes()
+		public delegate T TypeFromXmlReference<T>(string assemblyName, string clrNamespace, string typeName)
+			where T : class;
+
+		public static T GetTypeReference<T>(
```
FYI, this extension method is new as of yesterday because there would have otherwise been three separate places where this was being done almost identically . Same thing with the error message above.

legistek commented on Xamarin.Forms.Build.Tasks/XmlTypeExtensions.cs:
```diff
@@ -3,30 +3,49 @@
+			} else {
+				// Use standard XF assemblies
+				// (Should only happen in unit tests)
+				var requiredAssemblies = new[] {
```
I don't think this would ever happen in real life but this method was getting indirectly called in one of the unit tests with a dummy `module` that was assuming the standard assemblies were referenced.

legistek commented:
Hi @StephaneDelcroix  Sorry for all the git messages. It's ready for you now. I addressed your comments and also updated to incorporate what you merged in yesterday. Let me know if want me to change anything else.

legistek commented on Xamarin.Forms.Core/XmlnsDefinitionAttribute.cs:
```diff
@@ -2,11 +2,18 @@
 
 
-namespace Xamarin.Forms.Internals
+namespace Xamarin.Forms.Xaml
```
I don't think it is but I changed it back out of an abundance of caution

StephaneDelcroix commented:
build

StephaneDelcroix commented:
I rebased and squashed your work in this branch https://github.com/xamarin/Xamarin.Forms/tree/legistek-gh2691-xmlnsdefinitionattribute-public

legistek commented:
Oh I'm guessing that's the .NET Standard 1.x issue again. All the unit tests should probably be `#if`'ed to run only in .NET Standard 2.0 compilations because they will fail anyway even if they compile.

StephaneDelcroix commented:
@legistek I'm still on it. adding notes for myself and history

StephaneDelcroix commented:
Time added to XamlG at compilation is negligible.

StephaneDelcroix commented on Xamarin.Forms.Core/XmlnsDefinitionAttribute.cs:
```diff
@@ -6,7 +6,14 @@ namespace Xamarin.Forms.Internals
+	// so the attribute needs to remain internal in that case.
+	internal
+#endif
+		sealed class XmlnsDefinitionAttribute : Attribute
```
we have to change the namespace for this

StephaneDelcroix replied:
done

legistek commented:
@StephaneDelcroix done! I'll also add you to my repo in case you want to change anything else last minute.

legistek commented:
Hey @StephaneDelcroix, I was trying this out today on the Nuget preview and driving myself completely bonkers getting XamlC errors saying it couldn't find the class name in the custom namespace. Turns out I forgot to do the very thing I note is required in the PR description, namely to include a placeholder `Init` method in the control library that has to be called by the consuming library, otherwise .NET won't preserve the reference.

I noticed you removed my error message that warned about this (https://github.com/xamarin/Xamarin.Forms/commit/ec89923caa474e601cb4f6ab9cceb9fe63817ff7). Just thinking it might be worth either putting that back or at least in the documentation for `XmlnsDefinitionAttribute` adding a reminder about this.
